### PR TITLE
Backport #772 to release/rocm-5.4

### PR DESCRIPTION
**Pull request by jerryyin**
Revert "[hotfix] Disable conv when padding is an exact multiple of filter width (#770)"

This reverts commit 63d561b75386d528acb78e87a6188474c6e4cd12.

Resolve incorrect results by handling alignment in the autovectorizer

Previously, if we had, for example, an image with width 4 padded to width 6 (on the right) with a filter of 2, the vectorizer would infer a vectorization length of 2 in the M dimension because, all else being held constant, you could fetch the vectors (I[0], I[1]) (I[2], I[3]), [faild bounds check -> (0, 0)] without partial oob.

However, suppose we had added 1 to the input coordinate to handle the case where x = 1. Then we'd fetch the vectors (I[1], I[2]), (I[3], I[4]), (0, 0). Note that I[4], which is reading into the next row and should've been (I[3], 0) instead.

This commit resolves the issue by tracking alignment in the vectorizer, and using that information to resolve the vectorization of padding.

Add a test, loosen up alignment logic a bit to make it pass in a reasonable way

Fix up logic for merge, unmerge, embed

### CODE REVIEW

**Review by krzysz00 — APPROVED** [review]
Can we get the one last change (removing the `d.` edit) from the other PR?

**Comment by jerryyin** [comment]
Release testing:
 - xdlops: http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-773/3/pipeline/21/
 - nonxdlops: http://ml-ci.amd.com:21096/blue/organizations/jenkins/MLIR%2Frelease-branch-testing/detail/release-branch-testing/40/pipeline

**Review by yiqian1 — APPROVED** [review]

**Comment by jerryyin** [comment]
@krzysz00 
The existing condition is that nonxdlops release testing for this PR is passed whereas xdlops release testing is blocked by CI. I have tried kick off CI multiple times across the weekend but it either run into lockhart issues or blocked by limited instance of gfx908. I doubt the situation improve soon.

Do you think this PR will impact tuning in any way? I'm thinking of asking MIOpen to tune the 5.4 release MLIR package without this PR. Those configs that were blocked off are less likely to be production kernels, correct?

**Comment by krzysz00** [comment]
@jerryyin Since this PR doesn't impact the xdlops path, I claim we can fake the xdlops testing and merge

Now, as to tuning ... I'd want to basically look at the configs MIOpen tunes to see if they use filter width equal to the padding (where filter width > 1)

**Comment by jerryyin** [comment]
@krzysz00 Go ahead but please kick off a xdlops release testing on ROCm-5.4 branch. We need that record that there's a full pass in release testing.

@yiqian1 Please file the ticket to Tuna team to have them start tuning all solvers.

**Comment by krzysz00** [comment]
@yiqian1 Let's not file that ticket until this PR lands?

**Comment by yiqian1** [comment]
> @yiqian1 Let's not file that ticket until this PR lands?

Will file the ticket to Tuna team after this PR is in.

**Pull request merged**
